streamer/m3u8_concater.py

```diff
@@ -0,0 +1,955 @@
+                                    lang in langs
+                                    if len(codec_lang_division[codec][lang])]
+            sub_lang = MediaPlaylist._fit_missing_lang(aud_playlist_options,
+                                                       lang)
+            # Replace the empty codec_lang_division[codec][lang] with the
+            # substitution language with the same codec.
+            codec_lang_division[codec][lang] = codec_lang_division[codec][sub_lang]
+          # Sort the media playlists ascendingly based on the channel layouts.
+          codec_lang_division[codec][lang].sort(key=lambda pl: pl.channel_layout)
+          # Fill the division map for the current period from the codec_lang_division map.
+          for i, channel in enumerate(sorted(channels)):
+            division[codec][lang][channel].append(
+                codec_lang_division[codec][lang][min(
+                    i, len(codec_lang_division[codec][lang]) - 1)])
+    
+    return division
+  
+  @staticmethod
+  def concat_aud(all_aud_playlists: List[List['MediaPlaylist']]
+                 ) -> List['MediaPlaylist']:
+    """Concatenates multiple audio playlists into one multi-period playlist for
+    different languages, codecs, and channel layouts.
+    """
+    
+    division = MediaPlaylist.concat_aud_common(all_aud_playlists)
+    
+    concat_aud_playlists: List[MediaPlaylist] = []
+    for codec, lang_channel_div in division.items():
+      for lang, channel_div in lang_channel_div.items():
+        for channel, aud_playlists in channel_div.items():
+          stream_info = MediaPlaylist._similar_stream_info(
+              aud_playlists)
+          if lang != 'und':
+            # Put the language attribute in case it was removed.
+            stream_info['LANGUAGE'] = _quote(lang)
+          # Get a unique file name.
+          stream_info['NAME'], stream_info['URI'] = MediaPlaylist._next_unq_name()
+          # Get the max channels for this playlist.
+          stream_info['CHANNELS'] = MediaPlaylist._max_channels(aud_playlists)
+          concat_aud_playlist = MediaPlaylist(stream_info)
+          # Set the target duration.
+          concat_aud_playlist.target_duration = MediaPlaylist._max_targer_dur(
+              aud_playlists)
+          for aud_playlist in aud_playlists:
+            concat_aud_playlist.content += aud_playlist.content
+            # Add a discontinuity after each period.
+            concat_aud_playlist.content += '#EXT-X-DISCONTINUITY\n\n'
+          concat_aud_playlists.append(concat_aud_playlist)
+    
+    return concat_aud_playlists
+  
+  @staticmethod
+  def concat_aud_only(all_aud_playlists: List[List['MediaPlaylist']],
+                      all_var_playlists: List[List['MediaPlaylist']],
+                      durations: List[float]) -> List['MediaPlaylist']:
+    """Concatenates audio only periods with other audio only periods."""
+    
+    # Pair audio streams with their equivalent stream variants
```

Comment:
I honestly think it feels wrong that Packager spits out a "variant" that pairs with an audio group of itself, but I haven't spent the time digging through the spec to decide if it's required by the spec.  And Packager isn't the only one, and Player already has to handle it.  In any case, you've done well with this detail.